### PR TITLE
caffe 1.0 (new formula)

### PR DESCRIPTION
**Pull request by shoeffner**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
---> There is #22731 which I accidentally closed and decided to create a clean, new try from a proper feature branch without all the merge conflicts.
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Please see #22731 for details, I fixed all these problems:
- removed `head`
- no `make test` anymore
- in-tree build
- -j level is set automatically, no separate make command other than install
- no useless tests but a real smoke test for the command line tools
- Using `pkgshare` over `share` to install the resources
- Using `do` instead of `{...}` to split lines up into more readable bits in `resource(...).stage`

Additionally I set the Python build to OFF as it was not picked up by CMake, which I found out from the logs. Thus this formula only builds the caffe command line tools and C++ libraries to link against, but not the Matlab/Python bindings.

Sorry for the mess I created, I am still not sure what the best tactics is to rebase/merge the base branch into my PR branch to keep it up to date. I will just not touch it anymore, now that I used a proper feature branch, and read it up later.

### CODE REVIEW

**Comment by MindTooth** [comment]
According to this, I can see that some of the `depends_on` can be marked as optionals.

Docs: https://docs.brew.sh/Formula-Cookbook.html#specifying-other-formulae-as-dependencies

File: Formula/caffe.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    url "http://dl.caffe.berkeleyvision.org/bvlc_reference_caffenet.caffemodel"
+    sha256 "472d4a06035497b180636d8a82667129960371375bd10fcb6df5c6c7631f25e0"
+  end
+
+  def install
+    args = std_cmake_args + %w[
+      -DCPU_ONLY=ON
+      -DUSE_NCCL=OFF
+      -DBUILD_SHARED_LIBS=ON
+      -DBUILD_python=OFF
```
Comment (DomT4) [inline Formula/caffe.rb]: Is this correct? You've got a `python` hard dep further up.

**Comment by ilovezfs** [comment]
@shoeffner ping on @DomT4's question above.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

File: Formula/caffe.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  homepage "https://caffe.berkeleyvision.org/"
+  url "https://github.com/BVLC/caffe/archive/1.0.tar.gz"
+  sha256 "71d3c9eb8a183150f965a465824d01fe82826c22505f7aa314f700ace03fa77f"
+
+  depends_on "cmake" => :build
+  depends_on "boost"
+  depends_on "gflags"
+  depends_on "glog"
+  depends_on "hdf5"
+  depends_on "leveldb" => :optional
```
Comment (DomT4) [inline Formula/caffe.rb]: Deps should be alphabetised with `:build` options being first, and `:optional` options being last here. The one deviation from the normal rule is that `depends_on "snappy" if build.with? "leveldb"` should be the last dep here.

**Comment by fxcoudert** [comment]
@BrewTestBot test this please

**Review by fxcoudert — APPROVED** [review]

**Pull request closed without merging**